From 360a6adb956011ad81b112662ed60ca88fa725de Mon Sep 17 00:00:00 2001
From: Eike Ziller <eike.ziller@qt.io>
Date: Wed, 29 Apr 2026 11:14:26 +0200
Subject: [PATCH] Apply custom changes

- Remove XAR support, xar is an xml based format so it depends on a lot
  of libraries we don't want to include.
- Windows specific hacks
- Use GetTempPath2 where possible
---
 .../libarchive/libarchive/CMakeLists.txt      | 25 +++++++------------
 .../3rdparty/libarchive/libarchive/archive.h  |  7 +++---
 .../libarchive/libarchive/archive_entry.c     | 18 -------------
 .../libarchive/archive_entry_link_resolver.c  |  3 ---
 .../libarchive/archive_read_set_format.c      |  3 ---
 .../archive_read_support_format_all.c         |  8 +++---
 .../archive_read_support_format_by_code.c     |  2 --
 .../libarchive/libarchive/archive_util.c      | 16 ++++++++++--
 .../libarchive/archive_write_set_format.c     |  1 -
 .../archive_write_set_format_by_name.c        |  1 -
 10 files changed, 30 insertions(+), 54 deletions(-)

diff --git a/src/libs/3rdparty/libarchive/libarchive/CMakeLists.txt b/src/libs/3rdparty/libarchive/libarchive/CMakeLists.txt
index b29663742fd..1f585c3860a 100644
--- a/src/libs/3rdparty/libarchive/libarchive/CMakeLists.txt
+++ b/src/libs/3rdparty/libarchive/libarchive/CMakeLists.txt
@@ -107,12 +107,12 @@ SET(libarchive_SOURCES
   archive_read_support_format_iso9660.c
   archive_read_support_format_lha.c
   archive_read_support_format_mtree.c
-  archive_read_support_format_rar.c
+  # RAR has trouble on windows, so we disable it.
+  #archive_read_support_format_rar.c
   archive_read_support_format_rar5.c
   archive_read_support_format_raw.c
   archive_read_support_format_tar.c
   archive_read_support_format_warc.c
-  archive_read_support_format_xar.c
   archive_read_support_format_zip.c
   archive_string.c
   archive_string.h
@@ -166,7 +166,6 @@ SET(libarchive_SOURCES
   archive_write_set_format_ustar.c
   archive_write_set_format_v7tar.c
   archive_write_set_format_warc.c
-  archive_write_set_format_xar.c
   archive_write_set_format_zip.c
   archive_write_set_options.c
   archive_write_set_passphrase.c
@@ -255,20 +254,13 @@ IF(BUILD_SHARED_LIBS)
                         SOVERSION ${SOVERSION}
                         MACHO_COMPATIBILITY_VERSION ${MACHO_COMPATIBILITY_VERSION}
                         MACHO_CURRENT_VERSION ${MACHO_CURRENT_VERSION})
-  IF(WIN32 AND MSVC AND MSVC_USE_STATIC_CRT)
-    SET_PROPERTY(TARGET archive PROPERTY MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
-  ENDIF(WIN32 AND MSVC AND MSVC_USE_STATIC_CRT)
 ENDIF(BUILD_SHARED_LIBS)
 
 # archive_static is a static library
 ADD_LIBRARY(archive_static STATIC ${libarchive_SOURCES} ${include_HEADERS})
 TARGET_INCLUDE_DIRECTORIES(archive_static PUBLIC .)
 TARGET_LINK_LIBRARIES(archive_static ${ADDITIONAL_LIBS})
-SET_TARGET_PROPERTIES(archive_static PROPERTIES COMPILE_DEFINITIONS
-  LIBARCHIVE_STATIC)
-IF(WIN32 AND MSVC AND MSVC_USE_STATIC_CRT)
-  SET_PROPERTY(TARGET archive_static PROPERTY MSVC_RUNTIME_LIBRARY "MultiThreaded$<$<CONFIG:Debug>:Debug>")
-ENDIF(WIN32 AND MSVC AND MSVC_USE_STATIC_CRT)
+target_compile_definitions(archive_static PUBLIC LIBARCHIVE_STATIC)
 # On Posix systems, libarchive.so and libarchive.a can co-exist.
 IF(NOT WIN32 OR CYGWIN OR NOT BUILD_SHARED_LIBS)
   SET_TARGET_PROPERTIES(archive_static PROPERTIES OUTPUT_NAME archive)
@@ -279,15 +271,16 @@ IF(ENABLE_INSTALL)
   IF(BUILD_SHARED_LIBS)
     INSTALL(TARGETS archive
             RUNTIME DESTINATION bin
-            LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
-            ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
+            LIBRARY DESTINATION lib
+            ARCHIVE DESTINATION lib)
   ENDIF(BUILD_SHARED_LIBS)
   INSTALL(TARGETS archive_static
           RUNTIME DESTINATION bin
-          LIBRARY DESTINATION ${CMAKE_INSTALL_LIBDIR}
-          ARCHIVE DESTINATION ${CMAKE_INSTALL_LIBDIR})
+          LIBRARY DESTINATION lib
+          ARCHIVE DESTINATION lib)
   INSTALL_MAN(${libarchive_MANS})
   INSTALL(FILES ${include_HEADERS} DESTINATION include)
 ENDIF()
 
-add_subdirectory(test)
+
+set_property(TARGET archive_static PROPERTY POSITION_INDEPENDENT_CODE ON)
diff --git a/src/libs/3rdparty/libarchive/libarchive/archive.h b/src/libs/3rdparty/libarchive/libarchive/archive.h
index a7de6f6dd85..7bb8757de8c 100644
--- a/src/libs/3rdparty/libarchive/libarchive/archive.h
+++ b/src/libs/3rdparty/libarchive/libarchive/archive.h
@@ -378,8 +378,9 @@ typedef const char *archive_passphrase_callback(struct archive *,
 #define	ARCHIVE_FORMAT_AR_BSD			(ARCHIVE_FORMAT_AR | 2)
 #define	ARCHIVE_FORMAT_MTREE			0x80000
 #define	ARCHIVE_FORMAT_RAW			0x90000
-#define	ARCHIVE_FORMAT_XAR			0xA0000
-#define	ARCHIVE_FORMAT_LHA			0xB0000
+// Not supported since it needs xml and regex libraries we don't want to include
+//#define	ARCHIVE_FORMAT_XAR			0xA0000
+#define ARCHIVE_FORMAT_LHA 0xB0000
 #define	ARCHIVE_FORMAT_CAB			0xC0000
 #define	ARCHIVE_FORMAT_RAR			0xD0000
 #define	ARCHIVE_FORMAT_7ZIP			0xE0000
@@ -501,7 +502,6 @@ __LA_DECL int archive_read_support_format_rar5(struct archive *);
 __LA_DECL int archive_read_support_format_raw(struct archive *);
 __LA_DECL int archive_read_support_format_tar(struct archive *);
 __LA_DECL int archive_read_support_format_warc(struct archive *);
-__LA_DECL int archive_read_support_format_xar(struct archive *);
 /* archive_read_support_format_zip() enables both streamable and seekable
  * zip readers. */
 __LA_DECL int archive_read_support_format_zip(struct archive *);
@@ -864,7 +864,6 @@ __LA_DECL int archive_write_set_format_shar_dump(struct archive *);
 __LA_DECL int archive_write_set_format_ustar(struct archive *);
 __LA_DECL int archive_write_set_format_v7tar(struct archive *);
 __LA_DECL int archive_write_set_format_warc(struct archive *);
-__LA_DECL int archive_write_set_format_xar(struct archive *);
 __LA_DECL int archive_write_set_format_zip(struct archive *);
 __LA_DECL int archive_write_set_format_filter_by_ext(struct archive *a, const char *filename);
 __LA_DECL int archive_write_set_format_filter_by_ext_def(struct archive *a, const char *filename, const char * def_ext);
diff --git a/src/libs/3rdparty/libarchive/libarchive/archive_entry.c b/src/libs/3rdparty/libarchive/libarchive/archive_entry.c
index 5b8351b4a19..5a267a6893f 100644
--- a/src/libs/3rdparty/libarchive/libarchive/archive_entry.c
+++ b/src/libs/3rdparty/libarchive/libarchive/archive_entry.c
@@ -121,24 +121,6 @@ static const wchar_t	*ae_wcstofflags(const wchar_t *stringp,
 static const char	*ae_strtofflags(const char *stringp, size_t length,
 		    unsigned long *setp, unsigned long *clrp);
 
-#ifndef HAVE_WCSCPY
-static wchar_t * wcscpy(wchar_t *s1, const wchar_t *s2)
-{
-	wchar_t *dest = s1;
-	while ((*s1 = *s2) != L'\0')
-		++s1, ++s2;
-	return dest;
-}
-#endif
-#ifndef HAVE_WCSLEN
-static size_t wcslen(const wchar_t *s)
-{
-	const wchar_t *p = s;
-	while (*p != L'\0')
-		++p;
-	return p - s;
-}
-#endif
 #ifndef HAVE_WMEMCMP
 /* Good enough for simple equality testing, but not for sorting. */
 #define wmemcmp(a,b,i)  memcmp((a), (b), (i) * sizeof(wchar_t))
diff --git a/src/libs/3rdparty/libarchive/libarchive/archive_entry_link_resolver.c b/src/libs/3rdparty/libarchive/libarchive/archive_entry_link_resolver.c
index 947c65fca5f..c026534c256 100644
--- a/src/libs/3rdparty/libarchive/libarchive/archive_entry_link_resolver.c
+++ b/src/libs/3rdparty/libarchive/libarchive/archive_entry_link_resolver.c
@@ -147,9 +147,6 @@ archive_entry_linkresolver_set_strategy(struct archive_entry_linkresolver *res,
 	case ARCHIVE_FORMAT_ISO9660:
 	case ARCHIVE_FORMAT_SHAR:
 	case ARCHIVE_FORMAT_TAR:
-	case ARCHIVE_FORMAT_XAR:
-		res->strategy = ARCHIVE_ENTRY_LINKIFY_LIKE_TAR;
-		break;
 	default:
 		res->strategy = ARCHIVE_ENTRY_LINKIFY_LIKE_OLD_CPIO;
 		break;
diff --git a/src/libs/3rdparty/libarchive/libarchive/archive_read_set_format.c b/src/libs/3rdparty/libarchive/libarchive/archive_read_set_format.c
index 552ab12d2f9..bf33619c360 100644
--- a/src/libs/3rdparty/libarchive/libarchive/archive_read_set_format.c
+++ b/src/libs/3rdparty/libarchive/libarchive/archive_read_set_format.c
@@ -87,9 +87,6 @@ archive_read_set_format(struct archive *_a, int code)
     case ARCHIVE_FORMAT_WARC:
       str = "warc";
       break;
-    case ARCHIVE_FORMAT_XAR:
-      str = "xar";
-      break;
     case ARCHIVE_FORMAT_ZIP:
       str = "zip";
       break;
diff --git a/src/libs/3rdparty/libarchive/libarchive/archive_read_support_format_all.c b/src/libs/3rdparty/libarchive/libarchive/archive_read_support_format_all.c
index a2c5a4c6f54..8da47d006c1 100644
--- a/src/libs/3rdparty/libarchive/libarchive/archive_read_support_format_all.c
+++ b/src/libs/3rdparty/libarchive/libarchive/archive_read_support_format_all.c
@@ -59,7 +59,6 @@ archive_read_support_format_all(struct archive *a)
 	archive_read_support_format_lha(a);
 	archive_read_support_format_mtree(a);
 	archive_read_support_format_tar(a);
-	archive_read_support_format_xar(a);
 	archive_read_support_format_warc(a);
 
 	/*
@@ -70,9 +69,10 @@ archive_read_support_format_all(struct archive *a)
 	/* These have potentially large look-ahead. */
 	archive_read_support_format_7zip(a);
 	archive_read_support_format_cab(a);
-	archive_read_support_format_rar(a);
-	archive_read_support_format_rar5(a);
-	archive_read_support_format_iso9660(a);
+    // RAR does not compile on windows
+    //archive_read_support_format_rar(a);
+    archive_read_support_format_rar5(a);
+    archive_read_support_format_iso9660(a);
 	/* Seek is really bad, since it forces the read-ahead
 	 * logic to discard buffered data. */
 	archive_read_support_format_zip(a);
diff --git a/src/libs/3rdparty/libarchive/libarchive/archive_read_support_format_by_code.c b/src/libs/3rdparty/libarchive/libarchive/archive_read_support_format_by_code.c
index 982e2db70f9..c2299cc399f 100644
--- a/src/libs/3rdparty/libarchive/libarchive/archive_read_support_format_by_code.c
+++ b/src/libs/3rdparty/libarchive/libarchive/archive_read_support_format_by_code.c
@@ -65,8 +65,6 @@ archive_read_support_format_by_code(struct archive *a, int format_code)
 		return archive_read_support_format_tar(a);
 	case ARCHIVE_FORMAT_WARC:
 		return archive_read_support_format_warc(a);
-	case ARCHIVE_FORMAT_XAR:
-		return archive_read_support_format_xar(a);
 	case ARCHIVE_FORMAT_ZIP:
 		return archive_read_support_format_zip(a);
 	}
diff --git a/src/libs/3rdparty/libarchive/libarchive/archive_util.c b/src/libs/3rdparty/libarchive/libarchive/archive_util.c
index 5e0b916de43..3103f12449f 100644
--- a/src/libs/3rdparty/libarchive/libarchive/archive_util.c
+++ b/src/libs/3rdparty/libarchive/libarchive/archive_util.c
@@ -248,6 +248,18 @@ __archive_mktempx(const char *tmpdir, wchar_t *template)
 	ws = NULL;
 	archive_string_init(&temp_name);
 
+	typedef DWORD(WINAPI * GetTempPathFuncPtr)(DWORD, LPWSTR);
+
+	// We try to resolve GetTempPath2 and use that, otherwise fall back to GetTempPath:
+	static GetTempPathFuncPtr getTempPathW = 0;
+	if (!getTempPathW) {
+		const HMODULE kernel32 = GetModuleHandleW(L"kernel32.dll");
+		getTempPathW = (GetTempPathFuncPtr) GetProcAddress(kernel32, "GetTempPath2W");
+		if (!getTempPathW) {
+			getTempPathW = GetTempPathW;
+		}
+	}
+
 	if (template == NULL) {
 		/* Get a temporary directory. */
 		if (tmpdir == NULL) {
@@ -256,7 +268,7 @@ __archive_mktempx(const char *tmpdir, wchar_t *template)
 			size_t l, s;
 
 			s = MAX_PATH + 1;
-			l = GetTempPathW((DWORD)s, buf);
+			l = getTempPathW((DWORD)s, buf);
 			if (l == 0) {
 				la_dosmaperr(GetLastError());
 				goto exit_tmpfile;
@@ -272,7 +284,7 @@ __archive_mktempx(const char *tmpdir, wchar_t *template)
 					goto exit_tmpfile;
 				}
 				p = buf2 = tmp;
-				l = GetTempPathW((DWORD)s, buf2);
+				l = getTempPathW((DWORD)s, buf2);
 				if (l == 0) {
 					free(buf2);
 					la_dosmaperr(GetLastError());
diff --git a/src/libs/3rdparty/libarchive/libarchive/archive_write_set_format.c b/src/libs/3rdparty/libarchive/libarchive/archive_write_set_format.c
index f636cff7471..1651db0d9b4 100644
--- a/src/libs/3rdparty/libarchive/libarchive/archive_write_set_format.c
+++ b/src/libs/3rdparty/libarchive/libarchive/archive_write_set_format.c
@@ -60,7 +60,6 @@ struct { int code; int (*setter)(struct archive *); } codes[] =
 				archive_write_set_format_pax_restricted },
 	{ ARCHIVE_FORMAT_TAR_USTAR,	archive_write_set_format_ustar },
 	{ ARCHIVE_FORMAT_WARC,		archive_write_set_format_warc },
-	{ ARCHIVE_FORMAT_XAR,		archive_write_set_format_xar },
 	{ ARCHIVE_FORMAT_ZIP,		archive_write_set_format_zip },
 	{ 0,		NULL }
 };
diff --git a/src/libs/3rdparty/libarchive/libarchive/archive_write_set_format_by_name.c b/src/libs/3rdparty/libarchive/libarchive/archive_write_set_format_by_name.c
index 09519b12389..4eb63e066fc 100644
--- a/src/libs/3rdparty/libarchive/libarchive/archive_write_set_format_by_name.c
+++ b/src/libs/3rdparty/libarchive/libarchive/archive_write_set_format_by_name.c
@@ -72,7 +72,6 @@ struct { const char *name; int (*setter)(struct archive *); } names[] =
 	{ "v7tar",	archive_write_set_format_v7tar },
 	{ "v7",		archive_write_set_format_v7tar },
 	{ "warc",	archive_write_set_format_warc },
-	{ "xar",	archive_write_set_format_xar },
 	{ "zip",	archive_write_set_format_zip },
 	{ NULL,		NULL }
 };
-- 
2.53.0

